ipn/ipnlocal/serve.go

```diff
@@ -257,6 +257,96 @@ func (b *LocalBackend) ServeConfig() ipn.ServeConfigView {
 	return b.serveConfig
 }
 
+// StreamFunnel turns on Funnel for the given HostPort, and keeps it on
+// for the duration that the context is open.
+//
+// When the context is done, Funnel is turned off for the port.
+//
+// During the stream, any incoming connections to the port are reported
+// out to the provided io.Writer, in the format ipn.FunnelRequestLog.
+func (b *LocalBackend) StreamFunnel(ctx context.Context, w io.Writer, hp ipn.HostPort) (err error) {
+	port, err := hp.Port()
+	if err != nil {
+		return err
+	}
+
+	// Turn on Funnel for the given HostPort.
+	sc := b.ServeConfig().AsStruct()
+	mak.Set(&sc.AllowFunnel, hp, true)
+	if err := b.SetServeConfig(sc); err != nil {
+		return err
+	}
+
+	// Defer turning off Funnel once stream ends.
+	cleanup := func() {
+		sc := b.ServeConfig().AsStruct()
+		delete(sc.AllowFunnel, hp)
+		err = errors.Join(err, b.SetServeConfig(sc))
+	}
+	defer cleanup()
+
+	f, ok := w.(http.Flusher)
+	if !ok {
+		return errors.New("writer not a flusher")
+	}
+
+	var writeErrs []error
+	writeToStream := func(log ipn.FunnelRequestLog) {
+		jsonLog, err := json.Marshal(log)
+		if err != nil {
+			writeErrs = append(writeErrs, err)
+			return
+		}
+		if _, err := fmt.Fprintf(w, "%s\n", jsonLog); err != nil {
+			writeErrs = append(writeErrs, err)
+			return
+		}
+		f.Flush()
+	}
+
+	// Hook up connections stream.
+	b.mu.Lock()
+	mak.Set(&b.funnelStreamers, port, &writeToStream)
+	b.mu.Unlock()
+
+	select {
+	case <-ctx.Done():
```

Comment:
Comment on what should trigger `ctx.Done()`? Does this handle if the client process was killed (SIGKILL)?

Reply:
> Does this handle if the client process was killed (SIGKILL)?

Yeah it does, adding comment